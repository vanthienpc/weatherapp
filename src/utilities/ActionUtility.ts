import IAction from 'models/ActionModel';
import { call, put } from 'redux-saga/effects';
import ErrorResponseModel from 'models/ErrorResponseModel';

export function* createSagaEffect<P>(
  action: (...arg: any[]) => IAction<any>,
  effect: (...arg: any[]) => Promise<P | ErrorResponseModel>,
  ...args: any[]
): Generator {
  const response = yield call(effect, ...args);
  const isError: boolean = response instanceof ErrorResponseModel;

  return yield put(action(response, isError));
}

export const createAction = <T = undefined>(
  type: string,
  payload?: T,
  error = false,
  meta: any = null,
): IAction<T> => {
  return { type, payload, error, meta };
};
