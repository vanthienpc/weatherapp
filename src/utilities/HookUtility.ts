import {
  useSelector as originalUseSelector,
  useDispatch as originalUseDispatch,
  RootStateOrAny,
  TypedUseSelectorHook,
} from 'react-redux';

export const useSelector: TypedUseSelectorHook<RootStateOrAny> = (state: RootStateOrAny) =>
  originalUseSelector(state);
export const useDispatch = () => originalUseDispatch();
