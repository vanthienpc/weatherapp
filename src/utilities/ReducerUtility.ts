import { Reducer } from 'redux';
import IAction from 'models/ActionModel';

type ReducerMethod<T> = (state: T, action: IAction<any>) => T;
type ReducerMethods<T> = { [actionType: string]: ReducerMethod<T> };

export const handleReducer = <T = any>(initialState: T, methods: ReducerMethods<T>): Reducer<T> => {
  return (state: T = initialState, action: IAction<any>): T => {
    const method: ReducerMethod<T> | undefined = methods[action.type];

    if (!method || action.error) {
      return initialState;
    }

    return method(state, action);
  };
};
