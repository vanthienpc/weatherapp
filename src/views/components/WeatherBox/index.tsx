import React from 'react';
import { Row, Col, Card } from 'antd';
import { CaretUpOutlined, CaretDownOutlined } from '@ant-design/icons';
import styled from 'styled-components';
import IStore from 'models/IStore';
import { useSelector } from 'utilities/HookUtility';
import { WeatherModel } from 'stores/weather/models/WeatherModel';
import * as WeatherType from 'stores/weather/WeatherType';
import * as WeatherSelector from 'selectors/WeatherSelector';
import * as ErrorSelector from 'selectors/ErrorSelector';

const BoxContainer = styled(Row).attrs(() => ({
  gutter: 16,
}))`
  padding: 16px 0;
`;

const BoxCard = styled(Col).attrs(() => ({
  xs: 24,
  sm: 24,
  md: 8,
}))`
  text-align: center;
  margin-bottom: 16px;
`;

const FlexColumn = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
`;

const DayOfWeek = styled.span`
  display: block;
  font-size: 18px;
  font-weight: 700;
  margin-bottom: 16px;
`;

const Symbol = styled.img.attrs((props) => ({ ...props }))`
  margin-bottom: 16px;
`;

const TemperState = styled(FlexColumn)`
  font-weight: bold;
  text-transform: uppercase;
  margin-bottom: 30px;
`;

const TemperRange = styled.div`
  display: flex;
  justify-content: space-between;
`;

const TemperLimit = styled(FlexColumn)``;

const WeatherCard: React.FC<WeatherModel> = ({
  dayOfWeek,
  icon,
  state,
  temperCur,
  temperMin,
  temperMax,
}) => {
  return (
    <BoxCard>
      <Card hoverable>
        <DayOfWeek>{dayOfWeek}</DayOfWeek>
        <Symbol src={icon} loading="lazy" />
        <TemperState>
          {temperCur} <br /> {state}
        </TemperState>
        <TemperRange>
          <TemperLimit>
            <CaretUpOutlined />
            {temperMin}
            <span>Min</span>
          </TemperLimit>
          <TemperLimit>
            <CaretDownOutlined />
            {temperMax}
            <span>Max</span>
          </TemperLimit>
        </TemperRange>
      </Card>
    </BoxCard>
  );
};

const WeatherBox: React.FC = () => {
  const [weather, setWeather] = React.useState(undefined as WeatherModel[] | undefined);
  const selectError = useSelector((state: IStore) =>
    ErrorSelector.selectError(state, [WeatherType.FETCH_WEATHER_FINISHED]),
  );
  const selectWeatherListWithFiveDays = useSelector((state: IStore) =>
    WeatherSelector.selectWeatherListWithFiveDays(state),
  );

  React.useEffect(() => {
    setWeather(selectWeatherListWithFiveDays);
    if (Object.keys(selectError).length) {
      setWeather(undefined);
    }
  }, [selectWeatherListWithFiveDays, selectError]);

  if (weather) {
    return (
      <BoxContainer>
        {weather.map((item: WeatherModel) => (
          <WeatherCard key={item.id} {...item} />
        ))}
      </BoxContainer>
    );
  }

  return null;
};

export default WeatherBox;
