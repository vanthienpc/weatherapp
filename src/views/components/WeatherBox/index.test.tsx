import React from 'react';
import { Store } from 'redux';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import createSagaMiddleware from 'redux-saga';
import configureStore from 'redux-mock-store';
import * as HookUtility from 'utilities/HookUtility';
import IWeatherState from 'stores/weather/models/IWeatherState';
import WeatherBox from '.';

const setup = (initialState?: IWeatherState[] | undefined) => {
  let useEffect: any;
  let store: Store;

  const mockUseEffect = () => {
    useEffect.mockImplementationOnce((f: any) => f());
  };

  const sagaMiddleware = createSagaMiddleware();

  store = configureStore([sagaMiddleware])(initialState);

  useEffect = jest.spyOn(React, 'useEffect');
  mockUseEffect();
  mockUseEffect();

  jest.spyOn(HookUtility, 'useSelector').mockImplementation(() => store.getState());

  const setWeather = jest.fn();

  jest
    .spyOn(React, 'useState')
    .mockImplementation(() => [initialState ? store.getState() : undefined, setWeather]);

  //@ts-ignore
  const wrapper = shallow(<WeatherBox store={store} />);

  return { wrapper };
};

it('render should match the snapshot', () => {
  const initialState: IWeatherState[] = [
    {
      id: 1,
      dayOfWeek: 'Tuesday',
      temperMin: `29 °C`,
      temperMax: `31 °C`,
    },
  ];
  const { wrapper } = setup(initialState);
  expect(toJson(wrapper)).toMatchSnapshot();
});

it('render should match null', () => {
  const { wrapper } = setup();
  expect(wrapper.html()).toBeNull();
});
