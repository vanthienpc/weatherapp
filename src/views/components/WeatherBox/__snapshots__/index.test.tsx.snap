// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render should match the snapshot 1`] = `
<Styled(Row)>
  <WeatherCard
    dayOfWeek="Tuesday"
    icon="https://openweathermap.org/img/wn/10n@2x.png"
    id={1}
    key="1"
    state="Rain"
    temperCur="29 °C"
    temperMax="31 °C"
    temperMin="29 °C"
  />
</Styled(Row)>
`;
